include/deal.II/grid/grid_tools.h

```diff
@@ -2142,6 +2162,453 @@ namespace GridTools
             }
         }
   }
+
+
+
+  namespace internal
+  {
+    namespace ProjectToObject
+    {
+      struct CrossDerivative
+      {
+        const unsigned int direction_0;
+        const unsigned int direction_1;
+
+        CrossDerivative(const unsigned int d0, const unsigned int d1);
+      };
+
+      inline
+      CrossDerivative::CrossDerivative(const unsigned int d0, const unsigned int d1)
+        :
+        direction_0 (d0),
+        direction_1 (d1)
+      {}
+
+
+
+      template <typename F>
+      inline
+      auto
+      centered_first_difference(const double  center,
+                                const double  step,
+                                const F      &f)
+      -> decltype(f(center) - f(center))
```

Comment:
Didn't you wish we could already use C++14? Then you didn't have to provide the return type explicitly...

Reply:
I would still prefer to write down the output type, even in that case, just so that it is clearer what the code does.

Reply:
Yes, I might agree -- being explicit is useful for so many reasons, including documentation and for letting the compiler help you find bugs.

Reply:
When C++11 came out I started writing `auto` everywhere but now I try to avoid it unless the type is really, really long (or appears elsewhere in the expression) for these reasons.